WearSpeakerSample/wear/src/main/java/com/example/android/wearable/speaker/SpeakerApp.kt

```diff
@@ -0,0 +1,223 @@
+import kotlinx.coroutines.async
+import kotlinx.coroutines.coroutineScope
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.selects.select
+import kotlinx.coroutines.suspendCancellableCoroutine
+import kotlin.coroutines.resume
+
+/**
+ * The logic for the speaker sample.
+ *
+ * The stateful logic is kept by a [MainStateHolder].
+ */
+@Composable
+fun SpeakerApp() {
+    MaterialTheme {
+        lateinit var requestPermissionLauncher: ManagedActivityResultLauncher<String, Boolean>
+
+        val context = LocalContext.current
+        val activity = context.findActivity()
+        val scope = rememberCoroutineScope()
+
+        val stateHolder = remember(activity) {
+            MainStateHolder(
+                activity = activity,
+                requestPermission = {
+                    requestPermissionLauncher.launch(Manifest.permission.RECORD_AUDIO)
+                },
+                showPermissionRationale = {
+                    // TODO: Refactor away from normal AlertDialog to a Compose-specific dialog
+                    AlertDialog.Builder(activity)
+                        .setMessage(R.string.rationale_for_microphone_permission)
+                        .setPositiveButton(R.string.ok) { _, _ ->
+                            requestPermissionLauncher.launch(Manifest.permission.RECORD_AUDIO)
+                        }
+                        .setNegativeButton(R.string.cancel) { _, _ -> }
+                        .create()
+                        .show()
+                },
+                showSpeakerNotSupported = {
+                    AlertDialog.Builder(activity)
+                        .setMessage(R.string.no_speaker_supported)
+                        .setPositiveButton(R.string.ok) { _, _ -> }
+                        .create()
+                        .show()
+                }
+            )
+        }
+
+        val lifecycleOwner = LocalLifecycleOwner.current
+
+        requestPermissionLauncher = rememberLauncherForActivityResult(RequestPermission()) {
+            // We ignore the direct result here, since we're going to check anyway.
+            scope.launch {
+                lifecycleOwner.lifecycle.withStateAtLeast(Lifecycle.State.STARTED) {
+                    stateHolder.onAction(AppAction.PermissionResultReturned)
+                }
+            }
+        }
+
+        // Notify the state holder whenever we become started to reset the state
+        LaunchedEffect(stateHolder) {
+            lifecycleOwner.repeatOnLifecycle(Lifecycle.State.STARTED) {
+                stateHolder.onAction(AppAction.Started)
+            }
+        }
+
+        SpeakerScreen(
+            appState = stateHolder.appState,
+            isPermissionDenied = stateHolder.isPermissionDenied,
+            recordingProgress = stateHolder.recordingProgress,
+            onMicClicked = {
+                scope.launch {
+                    lifecycleOwner.lifecycle.withStateAtLeast(Lifecycle.State.STARTED) {
```

Comment:
These `withStateAtLeast` calls seem unnecessary since when buttons are clicked the app is supposed to be STARTED. Even if the `onAction` hasn't been processed. You might not want to delay and the app moves to the background, you might not want to delay this until the app goes back to the foreground